File: android/app/src/main/kotlin/com/example/flashfloodcommunication/communication/core/AlertCodec.kt
package com.example.flashfloodcommunication.communication.core

import java.nio.ByteBuffer
import java.nio.ByteOrder

object AlertCodec {

    // Pre-mapped instruction dictionary available locally offline on all devices
    val PRESET_INSTRUCTIONS = mapOf(
        1 to Pair("Evacuate immediately to higher ground!", "Turant unchi jagah par jayein!"),
        2 to Pair("Flash flood warning! Move away from river banks!", "Baadh ki chetavani! Nadi ke kinare se door rahein!"),
        3 to Pair("Emergency alert: Stay indoors at high altitude.", "Aapaatkaal: Unche sthan par andar hi rahein.")
    )

    /**
     * Stable 32-bit id used on the wire. Decoded packets use "ALERT_" + hex(hash);
     * re-encoding that form must produce the same hash or Device B relays mutate
     * into a new alert and bounce forever.
     */
    fun wireIdHash(alertId: String): Int {
        if (alertId.startsWith("ALERT_")) {
            val hex = alertId.substring(6)
            if (hex.isNotEmpty()) {
                try {
                    return hex.toLong(16).toInt()
                } catch (_: NumberFormatException) {
                    // Fall through to Java hashCode
                }
            }
        }
        return alertId.hashCode()
    }

    fun canonicalAlertId(alertId: String): String {
        return "ALERT_" + Integer.toHexString(wireIdHash(alertId)).uppercase()
    }

    fun encode(alert: AlertPacket): ByteArray {
        val buffer = ByteBuffer.allocate(16).order(ByteOrder.BIG_ENDIAN)

        // 1. Alert ID Hash (4 bytes) — keep stable across hop encode/decode
        buffer.putInt(wireIdHash(alert.alertId))

        // 2. Event Type (high 4 bits) + Severity (low 4 bits) -> 1 byte
        val eventCode = when (alert.eventType) {
            "FLASH_FLOOD" -> 1
            "DAM_BURST" -> 2
            else -> 0
        } and 0x0F
        val sevCode = alert.severity and 0x0F
        buffer.put(((eventCode shl 4) or sevCode).toByte())

        // 3. Risk Score (1 byte: 0-100)
        buffer.put(alert.riskScore.toInt().coerceIn(0, 100).toByte())

        // 4. Latitude (4 bytes: scaled by 1e6)
        buffer.putInt((alert.latitude * 1_000_000).toInt())

        // 5. Longitude (4 bytes: scaled by 1e6)
        buffer.putInt((alert.longitude * 1_000_000).toInt())

        // 6. TTL (high 4 bits) + Hop Count (low 4 bits) -> 1 byte
        val ttlNibble = alert.ttl.coerceIn(0, 15) and 0x0F
        val hopNibble = alert.hopCount.coerceIn(0, 15) and 0x0F
        buffer.put(((ttlNibble shl 4) or hopNibble).toByte())

        // 7. Instruction Codebook ID (1 byte)
        val instructionId = 1.toByte() // Defaults to Preset #1
        buffer.put(instructionId)

        return buffer.array()
    }

    fun decode(bytes: ByteArray): AlertPacket? {
        if (bytes.size < 16) return null

        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)

        // 1. Recover Alert ID
        val idHash = buffer.int
        val alertId = "ALERT_" + Integer.toHexString(idHash).uppercase()

        // 2. Event + Severity
        val eventAndSev = buffer.get().toInt()
        val eventCode = (eventAndSev ushr 4) and 0x0F
        val severity = eventAndSev and 0x0F
        val eventType = when (eventCode) {
            1 -> "FLASH_FLOOD"
            2 -> "DAM_BURST"
            else -> "EMERGENCY"
        }

        // 3. Risk Score
        val riskScore = (buffer.get().toInt() and 0xFF).toDouble()

        // 4. Lat / Lon
        val lat = buffer.int / 1_000_000.0
        val lon = buffer.int / 1_000_000.0

        // 5. TTL + Hop
        val ttlAndHop = buffer.get().toInt()
        val ttl = (ttlAndHop ushr 4) and 0x0F
        // Do not increment here — DtnManager bumps hopCount only when relaying
        val hopCount = ttlAndHop and 0x0F

        // 6. Instruction
        val instructionId = buffer.get().toInt() and 0xFF
        val instructions = PRESET_INSTRUCTIONS[instructionId] ?: Pair("Emergency flood alert!", "Aapaatkaal baadh chetavani!")

        val currentTime = System.currentTimeMillis()

        return AlertPacket(
            alertId = alertId,
            eventType = eventType,
            severity = severity,
            riskScore = riskScore,
            issuedAt = currentTime,
            expiresAt = currentTime + 3600000,
            latitude = lat,
            longitude = lon,
            instructionEn = instructions.first,
            instructionHi = instructions.second,
            priority = 10,
            ttl = ttl,
            hopCount = hopCount,
            copyCount = 1,
            payloadHash = "HASH_${Integer.toHexString(idHash)}",
            signature = "SIG_OFFLINE",
            originatorId = "BLE_MESH"
        )
    }
}

File: lib/communication/alert_codec.dart
import 'dart:typed_data';
import 'alert_packet.dart';

class AlertCodec {
  static const int maxHops = 3;

  static String canonicalAlertId(String id) {
    if (id.startsWith('ALERT_')) return id;
    return 'ALERT_${id.hashCode.toRadixString(16).toUpperCase()}';
  }

  static Uint8List encode(AlertPacket alert) {
    final bd = ByteData(16);
    // 1. Alert ID hash (4 bytes)
    bd.setInt32(0, alert.alertId.hashCode);
    // 2. Event code + severity (1 byte)
    final eventCode = (alert.eventType == 'FLASH_FLOOD' ? 1 : (alert.eventType == 'DAM_BURST' ? 2 : 3)) & 0x0F;
    final sevCode = alert.severity.clamp(0, 15) & 0x0F;
    bd.setUint8(4, (eventCode << 4) | sevCode);
    // 3. Risk score (1 byte)
    bd.setUint8(5, alert.riskScore.toInt().clamp(0, 100));
    // 4. Lat (4 bytes scaled 1e6)
    bd.setInt32(6, (alert.latitude * 1000000).toInt());
    // 5. Lon (4 bytes scaled 1e6)
    bd.setInt32(10, (alert.longitude * 1000000).toInt());
    // 6. TTL + Hop count (1 byte)
    final ttlNibble = alert.ttl.clamp(0, 15) & 0x0F;
    final hopNibble = alert.hopCount.clamp(0, 15) & 0x0F;
    bd.setUint8(14, (ttlNibble << 4) | hopNibble);
    // 7. Instruction ID (1 byte)
    bd.setUint8(15, 1);
    return bd.buffer.asUint8List();
  }

  static AlertPacket? decode(Uint8List bytes) {
    if (bytes.length < 16) return null;
    final bd = ByteData.sublistView(bytes);
    final idHash = bd.getInt32(0);
    final alertId = 'ALERT_${idHash.toRadixString(16).toUpperCase()}';
    final eventAndSev = bd.getUint8(4);
    final eventCode = (eventAndSev >> 4) & 0x0F;
    final severity = eventAndSev & 0x0F;
    final eventType = eventCode == 1 ? 'FLASH_FLOOD' : (eventCode == 2 ? 'DAM_BURST' : 'EMERGENCY_SOS');
    final riskScore = bd.getUint8(5).toDouble();
    final lat = bd.getInt32(6) / 1000000.0;
    final lon = bd.getInt32(10) / 1000000.0;
    final ttlAndHop = bd.getUint8(14);
    final ttl = (ttlAndHop >> 4) & 0x0F;
    final hopCount = ttlAndHop & 0x0F;

    return AlertPacket(
      alertId: alertId,
      eventType: eventType,
      severity: severity,
      riskScore: riskScore,
      issuedAt: DateTime.now().millisecondsSinceEpoch,
      expiresAt: DateTime.now().millisecondsSinceEpoch + 3600000,
      latitude: lat,
      longitude: lon,
      ttl: ttl,
      hopCount: hopCount,
      originatorId: 'BLE_PEER',
      message: 'Incoming Mesh SOS Signal',
    );
  }
}
